add "What is CernVM-FS?" section (#13)
multixscale/cvmfs-tutorial-hpc-best-practices: docs/00_what_is_cvmfs.md

of files that are served via outgoing HTTP connections only, thus avoiding problems with firewalls.
Internally, CernVM-FS uses [content-adressable storage (CAS)](https://en.wikipedia.org/wiki/Content-addressable_storage)
and [Merkle trees](https://en.wikipedia.org/wiki/Merkle_tree) (like [Git](https://git-scm.com/) also does)
to store file data and metadata.
Files in a CernVM-FS repository are *automatically downloaded* to a client system as
they are accessed, from web servers that support the CernVM-FS repository being used.

From an end user perspective, files in a CernVM-FS repository are available via subdirectory in `/cvmfs`,
with a user experience similar to that of an *on-demand streaming service* for music or video,
but then for software installations (or other types of data).

CernVM-FS was developed specifically for [distributing software](#distributing-software-as-main-use-case),
and provides several interesting features, including
[de-duplication of files](#de-duplication-of-files),
[compression of data](#compression-of-data),
[multi-level caching](#multi-level-caching),
and [verification of data integrity](#verification-of-data-integrity),

> and has been proven to scale to hundreds of millions of files and tens of thousands of clients.

Reviewer: ```suggestion
and has been proven to scale to billions of files and tens of thousands of clients.
```

## Use cases and features

### Distributing software as main use case

CernVM-FS was developed at [CERN](https://home.cern/) to let High Energy Physics (HEP) collaborations

> like the [Large Hadron Collider (LHC) project](https://home.cern/science/accelerators/large-hadron-collider)

Reviewer: ```suggestion
like the experiments at the [Large Hadron Collider (LHC)](https://home.cern/science/accelerators/large-hadron-collider)
```

> deploy software on the worldwide-distributed computing infrastructure used to run data processing applications,
> but can also be used to distribute large *data* repositories.

Reviewer: It may make sense to raise the data use case not too early. It may give the audience a wrong idea. Once needs to quite carefully look at the specific use case and perhaps build a dedicated infrastructure to make cvmfs work well with data repositories.

```suggestion
deploy software on the worldwide-distributed computing infrastructure used to run data processing applications.
In certain cases, the software has also been used to distribute large *data* repositories.
```

in a particular repository at different paths.

This can result in a significant reduction in storage capacity that is required to host a large stack of
many software installations, especially when identical files are spread out across the repository,
as often happens with particular files like example data files across multiple versions of the same software.

### Compression of data

CernVM-FS [stores file content with compression](https://cvmfs.readthedocs.io/en/stable/cpt-repo.html#compression-and-hash-algorithms),
on the server, which not only further reduces required storage space but also significantly limits the
network bandwidth that is used to serve that a stored files in a repository across the network.
On the client side, the data is transparently decompressed as a part of the POSIX-compatible way in which
the contents of a CernVM-FS repository is presented under `/cvmfs`.

### Automatic updating of repository contents

...

### Multi-level caching

To reduce the latency when accessing files in a CernVM-FS repository, a multi-level caching hierarchy
is employed.

First of all, CernVM-FS benefits from the standard Linux kernel file system cache on a client system:

> the content of a file that have been read recently will already be cached in memory by the kernel,
> so that file does not need to be read from disk again.

Reviewer: The contents and the meta-data. I.e. also most lookups are served from the kernel.
```suggestion
the content of a file that have been read recently will already be cached in memory by the kernel,
so that file does not need to be read from disk again.
```

on the client system with a configurable size, which uses a
[least-recently used (LRU)](https://en.wikipedia.org/wiki/Cache_replacement_policies#LRU) replacement policy.
Files in a CernVM-FS repository that have been accessed sufficiently recently on a particular system
will already be available in the local cache of that system, and do not need to be downloaded again from
the network of CernVM-FS servers that support that repository.

When the contents of a file that is being accessed is not available yet in the local client cache,
CernVM-FS will automatically (and transparently) download it either from a [forward proxy server](#squid-proxy)
(like [Squid](http://www.squid-cache.org/)), or from a [Stratum-1 replica server](#stratum-1-replica-server).
Both the proxy and the replica server could be within the same local network as the client, or not.
To help reduce performance problems regarding network latency and bandwidth, clients can leverage
the [Geo API](https://cvmfs.readthedocs.io/en/stable/cpt-replica.html#geo-api-setup) supported by
Stratum-1 replica servers to automatically sort them geographically,

> in order to prioritize connecting to the closest ones.

Reviewer: It may make sense to denote these caching levels as "site-level" / "cluster-level" cache and global mirror servers.

Furthermore, additional caches can be made available to CernVM-FS, such as an
[*alien cache*](https://cvmfs.readthedocs.io/en/stable/cpt-configure.html#alien-cache) on a shared cluster file system
like GPFS or Lustre that is not managed by CernVM-FS, and a
[Content Delivery Network (CDN)](https://en.wikipedia.org/wiki/Content_delivery_network)

> can be used to help limit the time required to download files that are not cached yet.

Reviewer: Perhaps these two aspects should be discussed separately. There is 

1. A special site cache deployment that uses the cluster file system as a replacement for the site's web caches. Particularly relevant for HPC
2. The possibility to use a (commercial) HTTP CDN such as Amazon S3 instead of own deployment web servers and proxies. This second point is not so much about caching but rather about turning a technical issue (running web servers and caches) into a commercial issue (paying a cloud provider).
